stream: use more explicit statements

BridgeAR opened the pull request:
Writing explicit statements produces better outcomes from the
compiler.
More important though: the instanceof Buffer check is expensive
and can be avoided as nothing besides strings could be passed
to the handler. This also fixes a TODO entry.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

##### Affected core subsystem(s)
streams

mscdex commented:
> Writing explicit statements produces better outcomes from the
> compiler.

Can you show how you determined this? Also, what is "better" ?

BridgeAR commented:
@mscdex this comes from @bmeurer (e.g. on a [meetup](https://www.youtube.com/watch?v=M1FBosB5tjM&feature=youtu.be&t=3348))

jasnell commented:
Ping @nodejs/streams

mscdex commented:
@BridgeAR That leaves a few questions though, like:

* What version(s) of V8 does that pertain to? 5.9? 6.0? 6.1? Is it something that is/can be backported to 5.9/6.0/etc.?

* Is this specific to `undefined`? Someone specifically asked about `null`, but it wasn't clear what the answer to that is/was or if this is an optimization that works for any explicit, strict equality/inequality check.

TimothyGu commented:
> nothing besides strings could be passed to the handler

How can you be sure of this? If you want to get rid of the `instanceof` check, `Object.getPrototypeOf() === Buffer.prototype` or `process.binding('util').isUint8Array` is generally faster.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8794/

BridgeAR commented:
@TimothyGu I forgot about the object mode even though it does not really work with the object mode and the error thrown in that situation is actually wrong. I added a test for the object mode and changed to throw that error always when using `objectMode`. Otherwise using objectMode would result in a weird situation where you can theoretically use this by only using `Buffer`s but nothing else.
Besides being in the `objectMode` the data event can only emit `string`s or `Buffer`s.

@mscdex there is more to read about this in his blog [post](http://benediktmeurer.de/2017/04/03/v8-behind-the-scenes-march-edition/). As far as I know it's mainly about Turbofan and that being explicit is always best but I guess @bmeurer can answer these questions better than me.

BridgeAR commented on lib/_stream_wrap.js:
```diff
@@ -47,12 +44,12 @@ function StreamWrap(stream) {
   });
   this.stream.on('data', function ondata(chunk) {
-    if (!(chunk instanceof Buffer)) {
+    if (typeof chunk === 'string' || this._readableState.objectMode === true) {
```
The object mode check could actually be moved in the `StreamWrap` constructor but it is currently possible to change the `objectMode` variable from an existing stream and in that case the upper check would not be sufficient anymore.
It is explicitly noted in the docs that this is not safe but I'm thinking about opening a PR to make the `objectMode` read only.

@mcollina do you think that's reasonable?

mcollina replied:
Streams implementors often change `objectMode` after the fact, so we cannot assume that. It'd a semver-major change, and several of us are -1 on those.

mcollina replied:
What you did here is ok.

BridgeAR commented:
PTAL

mcollina commented on lib/_stream_wrap.js:
```diff
@@ -47,12 +44,12 @@ function StreamWrap(stream) {
       this.removeListener('data', ondata);
 
-      self.emit('error', new errors.Error('ERR_STREAM_HAS_STRINGDECODER'));
+      self.emit('error', new errors.Error('ERR_STREAM_WRAP'));
```
Why this error code change?

BridgeAR replied:
Well, the error code was very specific and somewhat misleading that `objectMode` would or should not trigger this even though `objectMode` always would have resulted in a misleading error since it's first existence.

I think now it better represents both cases. I do see that it's not ideal to rename the code though.

jasnell replied:
Given that I don't believe the other error code had actually shipped in a release yet, this is fine. Definitely makes this semver-major tho.

jasnell replied:
The new error code needs to be included in the docs tho

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8842/

mcollina commented:
The linter is failing, can you fix that? `make lint` will show them.

mcollina commented:
I'm flagging this as semver-major because it changes the error code.

cc @jasnell

jasnell commented on lib/_stream_transform.js:
```diff
@@ -76,7 +76,7 @@ function afterTransform(er, data) {
   var cb = ts.writecb;
 
-  if (!cb) {
+  if (cb === null) {
```
This potentially could be a breaking change given that it would miss other falsy values potentially passed for `cb`

mcollina replied:
I checked. We are setting this internally, it's not user-driven.

BridgeAR commented:
Comment addressed and rebased due to a conflict.

@mcollina I could change the error in a separate commit to allow the main change without being semver-major if you like me to. But IMHO this is not semver-major as the error code has not yet landed in any release.

mcollina commented:
CI before landing: https://ci.nodejs.org/job/node-test-pull-request/8874/

mcollina commented:
Failures are unrelated, landing.

mcollina commented:
Landed as 1b54371c50cd6a67f22f6ec0db53bac3917387e5.
